blog/content/post/2020-08-19-system_latency.md

```diff
@@ -0,0 +1,172 @@
+Before we put the task to sleep lets remember current time
+
+1. The start point of measuring kernel latencies is when the interrupt occurs.
+2. The First thing kernel must do is observe it and while there is a lot to do and to observe even that part can take some time.
+3. After that, relevant ISR is called to handle interruption(in our case to wake up the task)
+4. Next goes the kernel scheduler whose job is to manage all of the processes working in the system. When our task has been woken up, it was placed on the CPU queue indeed. So now it's the scheduler thing to handle the task ASAP.
+5. When it's time come, CPU starts to process the task and we can end measuring the latency.
+
+After that, we can subtract the current time from the time before we put the task to sleep and get the latency by subtracting time difference with the time given to timer.
+
+The given example is the exact methodology of the [`cyclictest`](http://manpages.ubuntu.com/manpages/cosmic/man8/cyclictest.8.html) test program for testing system latencies which will be using here by us.
+
+### Additional load
+
+To simulate a stressful environment for the system we're testing tools like [`hackbench`](http://manpages.ubuntu.com/manpages/xenial/man8/hackbench.8.html) and [`stressapptest`](http://manpages.ubuntu.com/manpages/trusty/man1/stressapptest.1.html) can be used.
+
+* [`Hackbench`](http://manpages.ubuntu.com/manpages/xenial/man8/hackbench.8.html) - tool for stressing kernel scheduler by creating pairs of threads communicating with each other via sockets
+
+* [`Stressapptest`](http://manpages.ubuntu.com/manpages/trusty/man1/stressapptest.1.html) - program for generating a realistic load of memory, CPU, and I/O by creating a specified amount of threads writing to memory, to file, or communicate with given IP server
+
+## Testing platform
+
+Tests will be performed on an i.mx8 platform with two builds of yocto-linux. A Regular version and a specially configured as RT using rt patch.
+
+## Tests cases
+
+To get a complete system's characteristic at least a couple of tests with a different load must be performed. To get fully reliable results, a test's time must take as least a couple of hours.
+Although in our case to demonstrate results each test will take around 1 hour
+
+Test sheet:
+
+* **_CASE 1:_** `cyclictest` with no load
+
+* **_CASE 2:_** `cyclictest` with `hackbench` sending 128B data packages
+
+* **_CASE 3:_** `cyclictest` with `stressapptest` on 2 memory threads testing 256MB of memory
+
+* **_CASE 4:_** `cyclictest` with `stressapptest` on 4 memory threads testing 256MB of memory
+
+* **_CASE 5:_** `cyclictest` with `stressapptest` on 8 memory threads testing 256MB of memory, 4 I/O threads and 4 network threads
+
+## Results
+
+* **_CASE 1:_**(No load)
+_Regular Build_
+        <img src='../../static/img/system_latency_plots/normal/cyclic_alone.png' width=80%>
```

Comment:
I'm not sure if those links will render correctly in the blog. Please look at other posts (the path needs to be changed most likely).

Reply:
Should it be like `img/system_latency_plots/rt/cyclic_alone.png`?

Reply:
I believe the template post explains that: https://raw.githubusercontent.com/3mdeb/news-and-ideas/master/blog/content/post/YYYY-MM-DD-template-post.md

Reply:
Fixed